source/extensions/filters/http/grpc_json_reverse_transcoder/filter.h

```diff
@@ -0,0 +1,136 @@
+
+  Http::FilterHeadersStatus decodeHeaders(Http::RequestHeaderMap&, bool) override;
+
+  Http::FilterDataStatus decodeData(Buffer::Instance&, bool) override;
+
+  Http::FilterTrailersStatus decodeTrailers(Http::RequestTrailerMap&) override {
+    return Http::FilterTrailersStatus::Continue;
+  }
+
+  void setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks&) override;
+
+  Http::Filter1xxHeadersStatus encode1xxHeaders(Http::ResponseHeaderMap&) override {
+    return Http::Filter1xxHeadersStatus::Continue;
+  }
+
+  Http::FilterHeadersStatus encodeHeaders(Http::ResponseHeaderMap&, bool) override;
+
+  Http::FilterDataStatus encodeData(Buffer::Instance&, bool) override;
+
+  Http::FilterTrailersStatus encodeTrailers(Http::ResponseTrailerMap&) override;
+
+  Http::FilterMetadataStatus encodeMetadata(Http::MetadataMap&) override {
+    return Http::FilterMetadataStatus::Continue;
+  }
+
+  void setEncoderFilterCallbacks(Http::StreamEncoderFilterCallbacks&) override;
+
+  bool shouldTranscodeResponse() const { return transcoder_ != nullptr; }
+
+  // GrpcJsonReverseTranscoderFilter is neither copyable nor movable.
+  GrpcJsonReverseTranscoderFilter(const GrpcJsonReverseTranscoderFilter&) = delete;
+  GrpcJsonReverseTranscoderFilter& operator=(const GrpcJsonReverseTranscoderFilter&) = delete;
+
+private:
+  // MaybeExpandBufferLimits expands the buffer limits for the request and
+  // response if the limits are set in the reverse transcoder config and are
+  // greater than the default limits.
+  void MaybeExpandBufferLimits() const;
+  bool DecoderBufferLimitReached(uint64_t buffer_length) const;
+  bool EncoderBufferLimitReached(uint64_t buffer_length) const;
+  bool CheckAndRejectIfRequestTranscoderFailed() const;
+  bool CheckAndRejectIfResponseTranscoderFailed() const;
+  Grpc::Status::GrpcStatus GrpcStatusFromHeaders(Http::ResponseHeaderMap& headers);
+  void InitPerRouteConfig();
+  // BuildRequestFromHttpBody reads the contents of the data field of the
+  // google.api.HttpBody message and builds the request body out of it.
+  bool BuildRequestFromHttpBody(Buffer::Instance& data);
+  // SendHttpBodyResponse sends the response returned from the upstream server
+  // as a google.api.HttpBody message.
+  void SendHttpBodyResponse(Buffer::Instance* data);
+  void ReplaceAPIVersionInPath();
+  bool CreateDataBuffer(const nlohmann::json& payload, Buffer::OwnedImpl& buffer) const;
+  absl::Status ExtractHttpAnnotationValues(const Protobuf::MethodDescriptor* method_descriptor);
+
+  Http::StreamDecoderFilterCallbacks* decoder_callbacks_;
+  Http::StreamEncoderFilterCallbacks* encoder_callbacks_;
+
+  const GrpcJsonReverseTranscoderConfig* per_route_config_;
+  const std::shared_ptr<GrpcJsonReverseTranscoderConfig> config_;
+  std::unique_ptr<google::grpc::transcoding::Transcoder> transcoder_;
+  TranscoderInputStreamImpl request_in_;
+  TranscoderInputStreamImpl response_in_;
+
+  Buffer::OwnedImpl request_buffer_;
```

Comment:
You may want to expose counters for these buffers in the future, given that you have separate control over their sizes. Otherwise you limit visibility into memory usage.

Also this pattern will prevent Envoys overload manager from resetting most expensive streams under memory pressure, since it will not have visibility into these buffers. There is nothing you can do at this point about it, just pointing out why making memory usage of these buffers visible is important.

Reply:
Acknowledged!